protocol_grpc.go

```diff
@@ -795,19 +770,48 @@ func grpcParseTimeout(timeout string) (time.Duration, error) {
 	return time.Duration(num) * unit, nil
 }
 
-func grpcEncodeTimeout(timeout time.Duration) (string, error) {
+func grpcEncodeTimeout(timeout time.Duration) string {
+	div := func(d time.Duration, r time.Duration) int64 {
+		return int64(d / r) // round down
+	}
 	if timeout <= 0 {
-		return "0n", nil
+		return "0n"
 	}
-	for _, pair := range grpcTimeoutUnits {
-		digits := strconv.FormatInt(int64(timeout/pair.size), 10 /* base */)
-		if len(digits) < grpcMaxTimeoutChars {
-			return digits + string(pair.char), nil
-		}
+	if d := div(timeout, time.Nanosecond); d <= grpcTimeoutMaxValue {
+		return strconv.FormatInt(d, 10) + "n"
+	}
+	if d := div(timeout, time.Microsecond); d <= grpcTimeoutMaxValue {
+		return strconv.FormatInt(d, 10) + "u"
+	}
+	if d := div(timeout, time.Millisecond); d <= grpcTimeoutMaxValue {
+		return strconv.FormatInt(d, 10) + "m"
+	}
+	if d := div(timeout, time.Second); d <= grpcTimeoutMaxValue {
+		return strconv.FormatInt(d, 10) + "S"
+	}
+	if d := div(timeout, time.Minute); d <= grpcTimeoutMaxValue {
+		return strconv.FormatInt(d, 10) + "M"
+	}
+	return strconv.FormatInt(div(timeout, time.Hour), 10) + "H"
+}
+
+func grpcTimeoutUnitLookup(unit string) time.Duration {
```

Comment:
As a bit of a nit, we we explicitly use a `byte` here instead of a `string`, the switch statement gets optimized a bit more since the comparison is just with a uint8.

```
$ go test -v main_test.go -bench=. -benchmem
goos: darwin
goarch: arm64
BenchmarkString
BenchmarkString/hit
BenchmarkString/hit-10         	1000000000	         0.7803 ns/op	       0 B/op	       0 allocs/op
BenchmarkString/miss
BenchmarkString/miss-10        	768537562	         1.553 ns/op	       0 B/op	       0 allocs/op
BenchmarkByte
BenchmarkByte/hit
BenchmarkByte/hit-10           	1000000000	         0.3882 ns/op	       0 B/op	       0 allocs/op
BenchmarkByte/miss
BenchmarkByte/miss-10          	1000000000	         0.3111 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	command-line-arguments	3.126s
```

```go
// $ cat main_test.go
package main_test

import (
	"testing"
	"time"
)

var sink time.Duration

func grpcTimeoutUnitLookupString(unit string) time.Duration {
	switch unit {
	case "n":
		return time.Nanosecond
	case "u":
		return time.Microsecond
	case "m":
		return time.Millisecond
	case "S":
		return time.Second
	case "M":
		return time.Minute
	case "H":
		return time.Hour
	default:
		return 0
	}
}

func grpcTimeoutUnitLookupByte(unit byte) time.Duration {
	switch unit {
	case 'n':
		return time.Nanosecond
	case 'u':
		return time.Microsecond
	case 'm':
		return time.Millisecond
	case 'S':
		return time.Second
	case 'M':
		return time.Minute
	case 'H':
		return time.Hour
	default:
		return 0
	}
}

func BenchmarkString(b *testing.B) {
	b.Run("hit", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			sink = grpcTimeoutUnitLookupString("H")
		}
	})
	b.Run("miss", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			sink = grpcTimeoutUnitLookupString("x")
		}
	})
}

func BenchmarkByte(b *testing.B) {
	b.Run("hit", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			sink = grpcTimeoutUnitLookupByte('H')
		}
	})
	b.Run("miss", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			sink = grpcTimeoutUnitLookupByte('x')
		}
	})
}
```

https://godbolt.org/z/eh9arWedv

Reply:
You had me a `byte`. Switched!